util/progress/progressui/display.go

```diff
@@ -69,53 +88,208 @@ func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch
+	// TtyMode enforces the output is a tty and will otherwise cause an error if it isn't.
+	TtyMode DisplayMode = "tty"
+	// PlainMode is the human-readable plain text output. This mode is not meant to be read
+	// by machines.
+	PlainMode DisplayMode = "plain"
+	// RawJSONMode is the raw JSON text output. It will marshal the various solve status events
+	// to JSON to be read by an external program.
+	RawJSONMode DisplayMode = "raw:json"
+)
+
+// NewDisplay constructs a SolveStatusDisplay that outputs to the given console.File with the given DisplayMode.
+//
+// This method will return an error when the DisplayMode is invalid or if TtyMode is used but the console.File
+// does not refer to a tty. AutoMode will choose TtyMode or PlainMode depending on if the output is a tty or not.
+func NewDisplay(out console.File, mode DisplayMode, opts ...DisplaySolveStatusOpt) (SolveStatusDisplay, error) {
+	switch mode {
+	case AutoMode, TtyMode, DefaultMode:
+		if c, err := console.ConsoleFromFile(out); err == nil {
+			return NewConsoleDisplay(c, opts...), nil
+		} else if mode == "tty" {
+			return SolveStatusDisplay{}, errors.Wrap(err, "failed to get console")
+		}
+		fallthrough
+	case PlainMode:
+		return NewPlainDisplay(out, opts...), nil
+	case RawJSONMode:
+		return NewRawJSONDisplay(out, opts...), nil
+	default:
+		return SolveStatusDisplay{}, errors.Errorf("invalid progress mode %s", mode)
+	}
+}
+
+type consoleDisplay struct {
+	t              *trace
+	disp           *display
+	width, height  int
+	displayLimiter *rate.Limiter
+}
+
+// NewConsoleDisplay creates a new SolveStatusDisplay that prints a TTY
+// friendly output.
+func NewConsoleDisplay(c console.Console, opts ...DisplaySolveStatusOpt) SolveStatusDisplay {
+	dsso := newDisplaySolveStatusOpts(opts...)
+	if dsso.phase == "" {
+		dsso.phase = "Building"
+	}
+	return SolveStatusDisplay{
+		disp: &consoleDisplay{
+			t:    newTrace(c, true),
+			disp: &display{c: c, phase: dsso.phase, desc: dsso.consoleDesc},
+		},
+	}
+}
+
+func (d *consoleDisplay) init(displayTimeout time.Duration) {
+	d.displayLimiter = rate.NewLimiter(rate.Every(displayTimeout), 1)
+}
+
+func (d *consoleDisplay) update(ss *client.SolveStatus) {
+	d.width, d.height = d.disp.getSize()
+	d.t.update(ss, d.width)
+	if !d.displayLimiter.Allow() {
+		// Exit early as we are not allowed to update the display.
+		return
+	}
+	d.refresh()
+}
+
+func (d *consoleDisplay) refresh() {
+	d.disp.print(d.t.displayInfo(), d.width, d.height, false)
+}
+
+func (d *consoleDisplay) done() {
+	d.width, d.height = d.disp.getSize()
+	d.disp.print(d.t.displayInfo(), d.width, d.height, true)
+	d.t.printErrorLogs(d.t.w)
+}
+
+type plainDisplay struct {
+	t              *trace
+	printer        *textMux
+	displayLimiter *rate.Limiter
+}
+
+// NewPlainDisplay creates a new SolveStatusDisplay that outputs the status
+// in a human-readable plain-text format.
+func NewPlainDisplay(w io.Writer, opts ...DisplaySolveStatusOpt) SolveStatusDisplay {
```

Comment:
leave these private until someone has a use-case. `NewDisplay()` should be enough.

Reply:
I'll remove it. We had a section of code that manually used the plain display and it created this awkward code:

```
d, _ = progressui.NewDisplay(os.Stderr, progressui.PlainMode)
```

That's fine it just felt weird discarding the error. But, the error is needed for `NewDisplay` because tty mode can fail and the progress mode can be invalid.